alf/algorithms/sac_algorithm.py

```diff
@@ -224,63 +219,54 @@ def _actor_train_step(self, exp: Experience, state: SacActorState,
-                tape.watch(action)
-                critic1, critic1_state = self._critic_network1(
-                    critic_input,
-                    step_type=exp.step_type,
-                    network_state=state.critic1)
-
-                critic2, critic2_state = self._critic_network2(
-                    critic_input,
-                    step_type=exp.step_type,
-                    network_state=state.critic2)
+            critic1, critic1_state = self._critic_network1(
+                critic_input, state=state.critic1)
+            critic2, critic2_state = self._critic_network2(
+                critic_input, state=state.critic2)
 
-                target_q_value = tf.minimum(critic1, critic2)
+            target_q_value = torch.min(critic1, critic2)
 
-            dqda = tape.gradient(target_q_value, action)
+            dqda = torch.autograd.grad(target_q_value, action,
+                                       torch.ones(target_q_value.size()))[0]
 
             def actor_loss_fn(dqda, action):
                 if self._dqda_clipping:
-                    dqda = tf.clip_by_value(dqda, -self._dqda_clipping,
-                                            self._dqda_clipping)
+                    dqda = torch.clamp(dqda, -self._dqda_clipping,
+                                       self._dqda_clipping)
                 loss = 0.5 * losses.element_wise_squared_loss(
-                    tf.stop_gradient(dqda + action), action)
-                loss = tf.reduce_sum(
-                    loss, axis=list(range(1, len(loss.shape))))
+                    (dqda + action).detach(), action)
+                loss = loss.view(loss.size(0), -1).mean(1)
                 return loss
 
-            actor_loss = tf.nest.map_structure(actor_loss_fn, dqda, action)
-            alpha = tf.stop_gradient(tf.exp(self._log_alpha))
+            actor_loss = nest.map_structure(actor_loss_fn, dqda, action)
+            alpha = torch.exp(self._log_alpha).detach()
             actor_loss += alpha * log_pi
         else:
             critic1, critic1_state = self._critic_network1(
-                exp.observation,
-                step_type=exp.step_type,
-                network_state=state.critic1)
+                exp.observation, state=state.critic1)
 
             critic2, critic2_state = self._critic_network2(
-                exp.observation,
-                step_type=exp.step_type,
-                network_state=state.critic2)
+                exp.observation, state=state.critic2)
 
-            assert isinstance(action_distribution,
-                              tfp.distributions.Categorical), (
-                                  "Only `tfp.distributions.Categorical` " +
-                                  "was supported, received:" + str(
-                                      type(action_distribution)))
+            assert isinstance(
+                action_distribution, td.categorical.Categorical) or \
```

Comment:
use this instead: https://github.com/HorizonRobotics/alf/blob/4723afdba6329459cc16f3204ae901117c86aa05/alf/utils/summary_utils.py#L283

Reply:
Moved it to ```dist_utils.py``` as it is generally useful.